extensions/amp-auto-ads/0.1/ad-network.js

```diff
@@ -0,0 +1,98 @@
+/**
+ * Copyright 2016 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {DataAttributeDef} from './placement';
+import {buildUrl} from '../../../ads/google/a4a/url-builder';
+import {user} from '../../../src/log';
+
+/** @const */
+const TAG = 'amp-auto-ads';
+
+/**
+ * An interface intended to be implemented by any ad-networks wishing to support
+ * amp-auto-ads.
+ * @interface
+ */
+export class AdNetwork {
+
+  /**
+   * @return {string}
+   */
+  getConfigUrl() {}
+
+  /**
+   * Any data attributes derived from either the page or the auto-amp-ads tag
+   * that should be applied to any ads inserted.
+   * @return {!Array<!DataAttributeDef>}
+   */
+  getDataAttributes() {}
+}
+
+/**
+ * Builds and returns an AdNetwork instance for the given type.
+ * @param {string} type
+ * @param {!Element} autoAmpAdsElement
+ * @return {?AdNetwork}
+ */
+export function getAdNetwork(type, autoAmpAdsElement) {
+  if (type == 'adsense') {
+    return new AdSenseNetwork(autoAmpAdsElement);
+  }
+  user().error(TAG, 'No ad network implementation for type: ' + type);
+  return null;
+}
+
+/**
+ * @implements {AdNetwork}
+ */
+class AdSenseNetwork {
+  /**
+   * @param {!Element} autoAmpAdsElement
+   */
+  constructor(autoAmpAdsElement) {
+    this.autoAmpAdsElement_ = autoAmpAdsElement;
+  }
+
+  /**
+   * @override
+   */
+  getConfigUrl() {
+    return buildUrl('//pagead2.googlesyndication.com/getconfig/ama', [
```

Comment:
wondering if you want to cache the result.
do we expect this method to be called multiple times?